benchmark_v2/benchmark_framework.py

```diff
@@ -0,0 +1,1181 @@
+            except Exception as e:
+                if logger:
+                    logger.warning(f"Teardown callback failed for scenario {self.name}: {e}")
+    
+    def __repr__(self):
+        return f"BenchmarkScenario(name='{self.name}', variant='{self.config.variant}')"
+
+
+
+
+@dataclass
+class TimingResult:
+    """Result from a timing measurement."""
+    time_to_first_token: Optional[float] = None
+    latency: float = 0.0
+    tokens_per_second: Optional[float] = None
+    time_per_output_token_seconds: Optional[float] = None
+    total_tokens_generated: int = 0
+    metadata: Dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass
+class BenchmarkStatistics:
+    """Statistical analysis of benchmark measurements."""
+    name: str
+    measurements: List[float]
+    mean: float
+    median: float
+    std: float
+    min: float
+    max: float
+    p25: float  # 25th percentile
+    p75: float  # 75th percentile
+    p90: float  # 90th percentile
+    p95: float  # 95th percentile
+    p99: float  # 99th percentile
+    unit: str = "seconds"
+
+    @classmethod
+    def from_measurements(cls, name: str, measurements: List[float], unit: str = "seconds") -> 'BenchmarkStatistics':
+        """Create statistics from a list of measurements."""
+        if not measurements:
+            raise ValueError("Cannot create statistics from empty measurements")
+        
+        measurements_array = np.array(measurements)
+        
+        return cls(
+            name=name,
+            measurements=measurements,
+            mean=float(np.mean(measurements_array)),
+            median=float(np.median(measurements_array)),
+            std=float(np.std(measurements_array)),
+            min=float(np.min(measurements_array)),
+            max=float(np.max(measurements_array)),
+            p25=float(np.percentile(measurements_array, 25)),
+            p75=float(np.percentile(measurements_array, 75)),
+            p90=float(np.percentile(measurements_array, 90)),
+            p95=float(np.percentile(measurements_array, 95)),
+            p99=float(np.percentile(measurements_array, 99)),
+            unit=unit
+        )
+
+
+@dataclass 
+class HardwareInfo:
+    """Hardware information collected during benchmarking."""
+    gpu_name: str
+    gpu_memory_total: int  # MB
+    cpu_count: int
+    memory_total: int  # MB
+    python_version: str
+    torch_version: Optional[str] = None
+    cuda_version: Optional[str] = None
+
+
+@dataclass
+class BenchmarkMetadata:
+    """Metadata collected for each benchmark run."""
+    timestamp: str
+    commit_id: str
+    hardware_info: HardwareInfo
+    config: BenchmarkConfig
+
+
+class GPUMonitor:
+    """Monitor GPU utilization during benchmark execution."""
+    
+    def __init__(self, sample_interval: float = 0.1, logger: logging.Logger = None):
+        self.sample_interval = sample_interval
+        self.logger = logger or logging.getLogger(__name__)
+        self.monitoring = False
```

Comment:
Is using an [`Event`](https://docs.python.org/3/library/threading.html#event-objects) not preferable here?